### ISSUE TITLE
Automated Tagging of Text Data

### ISSUE THREAD

**Issue by ptitzlabs** [svoi-fr/mirai#48]
### Overview

The goal of this issue is to develop an algorithm that can automatically extract tags from text data, including both scraped documents and conversations on groups. The algorithm should be able to handle a wide range of topics related to immigration, asylum, and integration.

### Example Inputs and Outputs

The following are some examples of the inputs and outputs that the algorithm should be able to handle:

**Input:**
```css
from_id: user1360543352
id: 14
reply_to_message_id: 10
text: Конечно интересно, сегодня одна из немногих возможностей уехать из мордора
в нормальную цивилизованную страну это полит.убежище. Особенно если у вас есть
основания для этого )
type: answer
```
**Output:**
```json
{
  "message_id": 14,
  "user": 352,
  "chain": 10,
  "type": "conversation",
  "intent": ["answer", "misc"],
  "tags": ["asylum", "political_asylum_process_general"],
  "entities": [],
  "role": "user"
}
```
**Input:**
```css
from_id: user1360543352
id: 15
reply_to_message_id: 2
text: Я почему-то всегда считал что для получения даже краткосрочной тур.визы во
Францию нужно показывать на счету в банке как минимум сумму в 10 тыс.евро. Оказывается
даже имея 60 т.р. можно рассчитывать на её получение ))
type: statement
```
**Output:**
```json
{
  "message_id": 15,
  "user": 352,
  "chain": 2,
  "type": "conversation",
  "intent": ["share", "experience"],
  "tags": ["travel_visa"],
  "entities": [],
  "role": "user"
}
```
**Input:**
```css
from_id: user1360543352
id: 17
reply_to_message_id: 7
text: Очень даже неплохо, за бесплатно просто супер )
type: statement
```
**Output:**
```json
{
  "message_id": 17,
  "user": 352,
  "chain": 7,
  "type": "conversation",
  "intent": ["share", "opinion"],
  "tags": ["food_assistance"],
  "entities": [],
  "role": "user"
}
```
**Input:**
```css
from_id: channel1696664518
id: 18
reply_to_message_id: 7
text: Сначала было немного стыдно брать, а потом посчитали, мы евро 50 сэкономили!
Да, немного выкинули, но и осталось тоже не мало)
type: statement
```
**Output:**
```json
{
  "message_id": 18,
  "user": 518,
  "chain": 7,
  "type": "conversation",
  "intent": ["share", "experience"],
  "tags": ["food_assistance"],
  "entities": [],
  "role": "user"
}
```

### Data Collection and Labeling

Gemini AI and Google are both excellent tools for producing training datasets for this sort of thing, especially from live conversations. The first step in this issue is to collect and label a dataset of text data, including both scraped documents and conversations on groups.

The following is a list of some representative tags that the algorithm should be able to extract from the text, but this list is not exhaustive:

* appealing\_immigration\_court\_decision
* asylum
* asylum\_at\_the\_airport
* asylum\_case
* asylum\_case\_evidence
* asylum\_case\_third\_countries
* asylum\_eligibility
* asylum\_reasons
* asylum\_story\_preparation
* avoiding\_dublin\_regulation
* blocked\_bank\_accounts
* blocked\_websites
* choosing\_country
* crossing\_border
* detention
* eligibility\_for\_asylum
* financial\_aid
* find\_a\_place\_to\_rest
* find\_accommodation
* find\_accommodation\_urgent
* find\_accommodation\_shortterm
* find\_accommodation\_longterm
* food\_assistance
* free\_transport
* healthcare
* host\_country\_positive\_opinions
* housing\_procedures
* illegal\_crossing
* justify\_finances
* language\_barrier
* learn\_language
* legal\_procedures
* living\_costs
* misc
* money\_exchange
* organizations
* pensions
* political\_asylum\_process\_general
* preparing\_for\_departure
* route\_description
* route\_planning
* scams
* social\_benefits
* support\_programs
* translation\_services
* travel\_visa
* travelling\_with\_pets
* understanding\_the\_dublin\_regulation
* use\_vpn
* work\_permits

### Algorithm Development

Once the dataset has been collected and labeled, the next step is to develop an algorithm that can automatically extract tags from the text. This will involve designing the prompts, splitting the data into training and test sets, and training the algorithm to deduce the tags from the text or from the conversation.

The algorithm should be able to handle cases where the message is unrelated to anything and should be marked as null. The chain signifies the message chain, ideally it should tag "root" messages of the conversation, or any followup ones. This will be useful for creating conversation datasets to correlate with the associated flags.

### Evaluation and Deployment

The final step in this issue is to evaluate the performance of the algorithm and deploy it to a production environment. The evaluation should include metrics such as precision, recall, and F1 score, as well as a qualitative analysis of the algorithm's output.

Once the algorithm has been evaluated and any necessary improvements have been made, it can be deployed to a production environment, such as a web application or a chatbot, to automatically extract tags from text data and improve the user experience.
